content-src/components/DiscoveryStreamComponents/DSCard/DSCard.jsx

```diff
@@ -1,12 +1,69 @@
-import {actionCreators as ac} from "common/Actions.jsm";
+import {actionCreators as ac, actionTypes as at} from "common/Actions.jsm";
 import React from "react";
 
+const VISIBLE = "visible";
+const VISIBILITY_CHANGE_EVENT = "visibilitychange";
+const INTERSECTION_RATIO = 0.5;
+
 export class DSCard extends React.PureComponent {
   constructor(props) {
     super(props);
+
+    this.setupIntersectionObserver = this.setupIntersectionObserver.bind(this);
+    this.onIntersectionObserve = this.onIntersectionObserve.bind(this);
+    this.isIntersecting = this.isIntersecting.bind(this);
+    this.cardElementRef = this.cardElementRef.bind(this);
     this.onLinkClick = this.onLinkClick.bind(this);
   }
 
+  componentDidMount() {
+    if (this.props.document.visibilityState === VISIBLE) {
+      this.setupIntersectionObserver();
+    } else {
+      this._onVisibilityChange = () => {
+        if (this.props.document.visibilityState === VISIBLE) {
+          this.setupIntersectionObserver();
+          this.props.document.removeEventListener(VISIBILITY_CHANGE_EVENT, this._onVisibilityChange);
+        }
+      };
+      this.props.document.addEventListener(VISIBILITY_CHANGE_EVENT, this._onVisibilityChange);
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._onVisibilityChange) {
+      this.props.document.removeEventListener(VISIBILITY_CHANGE_EVENT, this._onVisibilityChange);
+    }
+    if (this._intersectionObserver) {
+      this._intersectionObserver.unobserve(this.cardElement);
+    }
+  }
+
+  setupIntersectionObserver() {
+    const options = {threshold: INTERSECTION_RATIO};
+    this._intersectionObserver = new IntersectionObserver(this.onIntersectionObserve, options);
+    this._intersectionObserver.observe(this.cardElement);
+  }
+
+  onIntersectionObserve(entries) {
+    for (let entry of entries) {
+      if (entry.isIntersecting && entry.intersectionRatio >= INTERSECTION_RATIO) {
+        this.isIntersecting();
```

Comment:
You might want to break this for loop once the expected intersection gets handled.